### PR TITLE
24142 ENS logging configuration setup

### PR DESCRIPTION
**Pull request by Siddartha-tech**
ENS Initial Logging and Configuration Setup

### CODE REVIEW

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs [hunk 1]
```diff
@@ -51,11 +84,69 @@ protected IConfigurationRoot BuildConfiguration(IWebHostEnvironment hostingEnvir
...
 
             builder.AddEnvironmentVariables();
+            var tempConfig = builder.Build();
+            string kvServiceUri = tempConfig["KeyVaultSettings:ServiceUri"];
+
+            if (!string.IsNullOrWhiteSpace(kvServiceUri))
+            {
+                builder.AddAzureKeyVault(new Uri(kvServiceUri),
+                    new DefaultAzureCredential(new DefaultAzureCredentialOptions { ManagedIdentityClientId = tempConfig["ESSManagedIdentity:ClientId"] }));
+            }
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs:94]: No such configuration added 'ESSManagedIdentity:ClientId'. Please update with correct configuration, if required.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
       "Microsoft.Hosting.Lifetime": "Information"
+    },
+        "ApplicationInsights": {
+            "LogLevel": {
+                "Default": "Information"
     }
   },
+        "PathFormat": "Logs/UKHO.External Notification Service.API-{Date}.txt",
+        "Json": true
+    },
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json:25]: "PathFormat": "Logs/UKHO.External Notification Service.API-{Date}.txt",
   "Json": true
Please remove these two properties from appsettings as these are not default properties for Logging.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        NotModified = 900006,
+        /// <summary>
+        /// 900007 - The requested resource responded with OK.
+        /// </summary>
+        OK = 900007,
+        /// <summary>
+        /// 900007 - The requested resource responded with Accepted.
+        /// </summary>
+        Accepted = 900008
+    }
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs:39]: Minor change to comment  as '/// 900008 - The requested resource responded with Accepted.' for Event Id Accepted = 900008

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/EventHubLoggingConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
```
Comment (usha13529) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/EventHubLoggingConfiguration.cs:7]: Remove unused usings.

**Review by usha13529 — COMMENTED** [review]
Reviewed and provided few minor review comments.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    {
+        private readonly IHttpContextAccessor httpContextAccessor;
+        protected readonly ILogger<T> Logger;
+        protected new HttpContext HttpContext => httpContextAccessor.HttpContext;
+        public const string LastModifiedDateHeaderKey = "Last-Modified";
+        public const string InternalServerError = "Internal Server Error";
+        public const string NotModified = "Not Modified";
+        protected string TokenAudience => httpContextAccessor.HttpContext.User.FindFirstValue("aud");
+        protected string TokenIssuer => httpContextAccessor.HttpContext.User.FindFirstValue("iss");
+
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs:22]: As of now i have not removed InternalServerError as it is used for default in **Switch** case and rest of them is removed.
Please suggest should I remove that too.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs (hunk 1 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs:94]: Yes. I missed that. Thanks for highlighting that.
Made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json (hunk 2 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json:27]: Yes. Removed from appsetting.json file.
Thanks!

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs (hunk 3 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Logging/EventIds.cs:39]: Maid changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/EventHubLoggingConfiguration.cs (hunk 4 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.Common/Configuration/EventHubLoggingConfiguration.cs:2]: Removed. Thanks!

**Review by barrie-cooper — COMMENTED** [review]
A number of comments relating to the coding guidelines suggested by the `.editorconfig` file that's in the root of the repo plus some missing unit test coverage for the majority of:
- UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs

These will need addressing please.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json [hunk 6]
```diff
@@ -1,10 +1,33 @@
...
-      "Microsoft.Hosting.Lifetime": "Information"
-    }
-  },
-  "AllowedHosts": "*"
+    "EventHubLoggingConfiguration": {
+        "MinimumLoggingLevel": "Warning",
+        "UkhoMinimumLoggingLevel": "Information",
+        "Environment": "dev",
+        "EntityPath": "",
+        "System": "ENS API",
```
Comment (barrie-cooper) [inline UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json:7]: Don't think it's an API:
```suggestion
        "System": "ENS",
```

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            var objectResult = new ObjectResult
+                (new InternalServerError
+                {
+                    CorrelationId = GetCurrentCorrelationId(),
+                    Detail = InternalServerError,
+                });
+            objectResult.StatusCode = StatusCodes.Status500InternalServerError;
+            return objectResult;
+        }
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs:57]: unit test cases while implementing for actual class. As BaseController is abstract class

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        private IActionResult BuildAcceptedResponse()
+        {
+            LogInfo(EventIds.Accepted.ToEventId(), "Accepted", GetCurrentCorrelationId());
+            return new StatusCodeResult(StatusCodes.Status202Accepted);
+        }
+
+        private void LogError(EventId eventId, List<Error> errors, string errorType, string correlationId)
+        {
+            Logger.LogError(eventId, $"{HttpContext.Request.Path} - {errorType} - {{Errors}} for CorrelationId - {{correlationId}}", errors, correlationId);
+        }
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs:98]: unit test cases while implementing for actual class. As BaseController is abstract class

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Filters/LoggingMiddleware.cs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+namespace UKHO.ExternalNotificationService.API.Filters
+{
+    [ExcludeFromCodeCoverage]   //Used in Startup.cs
+    public static class LoggingMiddleware
+    {
+        private const string RedactedValue = "********";
+        private static readonly string[] HeadersToRedact = { "userpass", "token" };
+        private const int maxBodyCharSize = 1000;
+        private const int truncatedBodyCharSize = 987;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Filters/LoggingMiddleware.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs [hunk 10]
```diff
@@ -51,11 +84,69 @@ protected IConfigurationRoot BuildConfiguration(IWebHostEnvironment hostingEnvir
...
+                        additionalValues["_Service"] = eventHubLoggingConfiguration.Value.Service;
+                        additionalValues["_NodeName"] = eventHubLoggingConfiguration.Value.NodeName;
+                        additionalValues["_RemoteIPAddress"] = httpContextAccessor.HttpContext.Connection.RemoteIpAddress.ToString();
+                        additionalValues["_User-Agent"] = httpContextAccessor.HttpContext.Request.Headers["User-Agent"].FirstOrDefault() ?? string.Empty;
+                        additionalValues["_AssemblyVersion"] = Assembly.GetExecutingAssembly().GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version;
+                        additionalValues["_X-Correlation-ID"] =
+                            httpContextAccessor.HttpContext.Request.Headers?[CorrelationIdMiddleware.XCorrelationIdHeaderKey].FirstOrDefault() ?? string.Empty;
+
+                        if (httpContextAccessor.HttpContext.User.Identity.IsAuthenticated)
+                            additionalValues["_UserId"] = httpContextAccessor.HttpContext.User.FindFirstValue("http://schemas.microsoft.com/identity/claims/objectidentifier");
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Startup.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs]: Removed

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace UKHO.ExternalNotificationService.API.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public abstract class BaseController<T> : ControllerBase
+    {
+        private readonly IHttpContextAccessor httpContextAccessor;
+        protected readonly ILogger<T> Logger;
+        protected new HttpContext HttpContext => httpContextAccessor.HttpContext;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        public const string InternalServerError = "Internal Server Error";
+
+        protected BaseController(IHttpContextAccessor httpContextAccessor, ILogger<T> logger)
+        {
+            this.httpContextAccessor = httpContextAccessor;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            Logger = logger;
+        }
+
+        protected string GetCurrentCorrelationId()
+        {
+            return httpContextAccessor.HttpContext.Request.Headers[CorrelationIdMiddleware.XCorrelationIdHeaderKey].FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/SubscriptionController.cs [hunk 15]
```diff
@@ -1,21 +1,29 @@
...
         {
+            _logger = logger;
         }
 
         [HttpPost]
-        public async Task<IActionResult> Post([FromBody] JObject jobj)
+        public async virtual Task<IActionResult> Post([FromBody] JObject jobj)
         {
-            return Ok();
+            _logger.LogInformation(EventIds.LogRequest.ToEventId(), "Subcription request Accepted", jobj);
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/SubscriptionController.cs]: correction made

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/SubscriptionController.cs [hunk 16]
```diff
@@ -1,21 +1,29 @@
 ﻿using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
 using Newtonsoft.Json.Linq;
+using System.Net;
+using UKHO.ExternalNotificationService.Common.Logging;
+using UKHO.ExternalNotificationService.Common.Models.Response;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/SubscriptionController.cs]: Yes sorted changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs [hunk 17]
```diff
@@ -10,11 +13,16 @@ namespace UKHO.ExternalNotificationService.API.UnitTests.Controllers
     public class SubscriptionControllerTests
     {
         private SubscriptionController _controller;
+        private ILogger<SubscriptionController> fakeLogger;
+        private IHttpContextAccessor fakeHttpContextAccessor;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs [hunk 18]
```diff
@@ -10,11 +13,16 @@ namespace UKHO.ExternalNotificationService.API.UnitTests.Controllers
...
 
         [SetUp]
         public void Setup()
         {
-            _controller = new SubscriptionController();
+            fakeHttpContextAccessor = A.Fake<IHttpContextAccessor>();
+            fakeLogger = A.Fake<ILogger<SubscriptionController>>();
+            A.CallTo(() => fakeHttpContextAccessor.HttpContext).Returns(new DefaultHttpContext());
+            _controller = new SubscriptionController(fakeHttpContextAccessor, fakeLogger);
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API.UnitTests/Controllers/SubscriptionControllerTests.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json (hunk 6 above)
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/appsettings.json]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+using UKHO.ExternalNotificationService.Common.Logging;
+using UKHO.ExternalNotificationService.Common.Models.Response;
+
+namespace UKHO.ExternalNotificationService.API.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public abstract class BaseController<T> : ControllerBase
+    {
+        private readonly IHttpContextAccessor httpContextAccessor;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs]: Yes made changes accordingly.

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+                    return BuildOkResponse(model);
+
+                case HttpStatusCode.InternalServerError:
+                    return BuildInternalServerErrorResponse();
+
+                case HttpStatusCode.BadRequest:
+                    return BuildBadRequestErrorResponse(errors);
+
+                default:
+                    return BuildInternalServerErrorResponse();
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Controllers/BaseController.cs:79]: unit test cases while implementing for actual class. As BaseController is abstract class
So, Excluded from code coverage

File: UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Filters/LoggingMiddleware.cs [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+                {
+                    if (bodyAsString.Contains(propertyToRedact))
+                    {
+                        bodyAsString = RedactBody(propertyToRedact, bodyAsString, logger);
+                    }
+                }
+
+                if (bodyAsString.Length > maxBodyCharSize)
+                {
+                    bodyAsString = bodyAsString.Remove(truncatedBodyCharSize) + "... Truncated";
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExternalNotificationService.API/UKHO.ExternalNotificationService.API/Filters/LoggingMiddleware.cs]: Yes made changes accordingly.

**Review by usha13529 — APPROVED** [review]
LGTM.

File: NVDSuppressions.xml [hunk 22]
```diff
@@ -9,4 +9,121 @@
...
+    <cve>CVE-2021-26423</cve>
+  </suppress>
+  <suppress>
+    <notes>
+      <![CDATA[
+    file name: Microsoft.AspNetCore.Hosting.Abstractions.dll
+    See https://github.com/dotnet/announcements/issues/178 for full MS Details on CVE.
+    A remote code execution vulnerability exists in .NET 5 and .NET Core due to how text encoding is performed.This vulnerability is for System.Text.Encodings.Web and fixed on 5.0.1 and we are using 5.0.1
+    This CVE applies to any .NET Core 3.1 but patch is provided at 3.1.407, ESS is pinned to .Net core SDK > 3.1.408 which is not vunerable, and deployed to Azure App Service which will have the security patch applied.
+    ]]>
```
Comment (usha13529) [inline NVDSuppressions.xml:70]: Please update suppression note as 'ENS is pinned to .Net core SDK > 6.0.101'

File: NVDSuppressions.xml [hunk 23]
```diff
@@ -9,4 +9,121 @@
...
+    <cve>CVE-2021-26701</cve>
+  </suppress>
+  <suppress>
+    <notes>
+      <![CDATA[
+    file name: Microsoft.AspNetCore.Hosting.Abstractions.dll
+    See https://github.com/dotnet/announcements/issues/196 for full MS Details on CVE.
+    An information disclosure vulnerability exists in .NET 5.0, .NET Core 3.1 and .NET Core 2.1 when dumps created by the tool to collect crash dumps and dumps on demand are created with global read permissions on Linux and macOS.
+    This CVE applies to any .NET Core 3.1 but patch is provided at 3.1.412, ESS is pinned to .Net core SDK > 3.1.408 which is not vunerable, and deployed to Azure App Service which will have the security patch applied.
+    ]]>
```
Comment (usha13529) [inline NVDSuppressions.xml:81]: Please update suppression note as 'ENS is pinned to .Net core SDK > 6.0.101'

File: NVDSuppressions.xml [hunk 24]
```diff
@@ -9,4 +9,121 @@
...
+  </suppress>
+  <suppress>
+    <notes>
+      <![CDATA[
+    file name: Microsoft.AspNetCore.Hosting.Abstractions.dll
+    See https://github.com/dotnet/announcements/issues/176 for full MS Details on CVE.
+    A remote code execution vulnerability exists when parsing certain types of graphics files.
+    This vulnerability only exists on systems running on MacOS or Linux.
+    This CVE applies to any .NET Core 3.1 but patch is provided at 3.1.405, ESS is pinned to .Net core SDK > 3.1.408 which is not vunerable, and deployed to Azure App Service which will have the security patch applied.
+    ]]>
```
Comment (usha13529) [inline NVDSuppressions.xml:93]: Please update suppression note as 'ENS is pinned to .Net core SDK > 6.0.101'

**Review by usha13529 — COMMENTED** [review]
Minor changes to NVD suppression notes please.

File: NVDSuppressions.xml (hunk 22 above)
Comment (Siddartha-tech) [reply NVDSuppressions.xml:70]: updated accordingly.

File: NVDSuppressions.xml (hunk 24 above)
Comment (Siddartha-tech) [reply NVDSuppressions.xml:93]: updated accordingly.

File: NVDSuppressions.xml (hunk 23 above)
Comment (Siddartha-tech) [reply NVDSuppressions.xml:81]: updated accordingly.

**Pull request merged**